### PR TITLE
Change `// test_output! output` to `//! test_output(output)`

### PR DESCRIPTION
**Pull request by SamantaTarun**
#7

### CODE REVIEW

**Comment by voltrevo** [comment]
Great to see a start here @tarunsamanta2k20.

I've marked the PR as a draft as the changes are incomplete. Let me know if you need any help completing #7.

**Comment by SamantaTarun** [comment]
Please let me know, if any, changes require.

File: vstc/src/test_inputs.rs [hunk 1]
```diff
@@ -32,11 +32,11 @@ mod tests {
       let file_contents = fs::read_to_string(&file_path).expect("Failed to read file contents");
 
       if let Some(first_line) = file_contents.lines().next() {
-        if first_line.starts_with("// test_output! ") {
+        if first_line.starts_with("//! test_output ") {
```
Comment (voltrevo) [inline vstc/src/test_inputs.rs]: This won't match the new format, the start of the pattern is `"//! test_output("`.

**Review by voltrevo — CHANGES_REQUESTED** [review]
Heya. Thanks for the update, more changes are needed though.

In addition to my comment:
- the `)` of the annotation needs to be handled
- all the files with `// test_output! ...` need to be changed, not just `anonymousMain.ts`

Use `cargo test` locally to check your changes. Also sample a few of the annotations and check that they are actually being tested by (temporarily) adjusting the annotation to ensure the test breaks when it asserts the wrong output.

If you have discord it might be better to discuss in more detail over there. I'm voltrevo#6526.

File: vstc/src/test_inputs.rs (hunk 1 above)
Comment (voltrevo) [reply vstc/src/test_inputs.rs]: Optional: I'd also suggest that we filter on just `//!`, and then if the pattern isn't recognized, we fail the file for that too.

**Comment by SamantaTarun** [comment]
@voltrevo let me give a try.

**Comment by SamantaTarun** [comment]
@voltrevo please let me know clearly what needed to be done?

**Comment by voltrevo** [comment]
@tarunsamanta2k20 The following is still needed:

1. All files containing `test_output!` need to be updated (about 100 files), not just the one you changed
2. It needs to work correctly. `if first_line.starts_with("//! test_output ") {` is wrong - it results in no tests being found.

I can see that `cargo test` technically passes on your branch, but this is because none of the tests are being found. Use `cargo test -p vstc -- --nocapture` to get a more verbose output. This is the output from the main branch:

<details>
<summary>Click to expand</summary>

```
➜  ValueScript git:(main) cargo test -p vstc -- --nocapture
   Compiling valuescript_common v0.1.0 (/Users/andrew/workspaces/ValueScript/valuescript_common)
   Compiling valuescript_compiler v0.1.0 (/Users/andrew/workspaces/ValueScript/valuescript_compiler)
   Compiling valuescript_vm v0.1.0 (/Users/andrew/workspaces/ValueScript/valuescript_vm)
   Compiling vstc v0.1.0 (/Users/andrew/workspaces/ValueScript/vstc)
    Finished test [unoptimized + debuginfo] target(s) in 5.08s
     Running unittests src/main.rs (target/debug/deps/vstc-1458690d5106a7a4)

running 1 test

Testing /Users/andrew/workspaces/ValueScript/inputs/failing/numberMethods/toExponentialInconsistent.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/failing/stringIteration.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/anonymousMain.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/arrayBuiltin.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/assignmentEvalOrder.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/assignmentEvalOrder2.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/bigint/noMix.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/binaryTree.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/boolean.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/capture.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/captureButNotRefBeforeInit.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/captureShadowed.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/complexPatternOrder.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/counter.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/destructuring/decl/array.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/destructuring/decl/multi_complex.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/destructuring/decl/object.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/destructuring/param/array.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/destructuring/param/multi_complex.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/destructuring/param/object.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/dontRevertMath.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/doublePreInc.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/exceptions/catchDestructure.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/exceptions/finallyBreakSwallowsException.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/exceptions/finallyBreakSwallowsReturn.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/exceptions/finallyReturnSwallowsException.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/exceptions/finallyWithoutThrow.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/exceptions/isReturningReset.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/exceptions/nestedFinallyReturn.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/exceptions/nestedTry.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/exceptions/revertMethodCallOnCatch.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/exceptions/revertOnCatch.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/exceptions/revertSmallQueue.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/exceptions/tryCatch.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/exceptions/tryCatchFinally.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/exceptions/tryFinally.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/exceptions/tryReturnFinallyReturn.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/forOf.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/hello.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/importBar.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/importFoo.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/importFooAndBar.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/importFoobar.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/mergeSort.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/multiAssignment.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/multiCompoundAssignment.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/mutatedRegisterIssue.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/nestedAssignment.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/nestedCompoundAssignment.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/nestedFn.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/nestedFnCapture.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/numberMethods/isInteger.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/numberMethods/isSafeInteger.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/numberMethods/parseFloat.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/numberMethods/parseInt.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/numberMethods/toExponential.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/numberMethods/toFixed.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/patternEvalOrder.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/patternEvalOrder2.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/preincrement.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/projEuler/p1.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/projEuler/p11.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/projEuler/p13.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/projEuler/p15.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/projEuler/p16.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/projEuler/p18.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/projEuler/p19.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/projEuler/p2.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/projEuler/p25.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/projEuler/p3.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/projEuler/p4.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/projEuler/p5.ts ...
Debug.log: { factors: [ 2, 2, 3 ] }
Debug.log: { factors: [ 2, 2, 3, 5 ] }
Debug.log: { factors: [ 2, 2, 2, 3, 5, 7 ] }
Debug.log: { factors: [ 2, 2, 2, 3, 3, 5, 7 ] }
Debug.log: { factors: [ 2, 2, 2, 3, 3, 5, 7 ] }
Debug.log: { factors: [ 2, 2, 2, 3, 3, 5, 7, 11 ] }
Debug.log: { factors: [ 2, 2, 2, 3, 3, 5, 7, 11, 13 ] }
Debug.log: { factors: [ 2, 2, 2, 3, 3, 5, 7, 11, 13 ] }
Debug.log: { factors: [ 2, 2, 2, 2, 3, 3, 5, 7, 11, 13 ] }
Debug.log: { factors: [ 2, 2, 2, 2, 3, 3, 5, 7, 11, 13, 17 ] }
Debug.log: { factors: [ 2, 2, 2, 2, 3, 3, 5, 7, 11, 13, 17, 19 ] }

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/projEuler/p6.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/projEuler/p8.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/projEuler/p9.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/readme-demo.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/recursiveCapture.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/shadow.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringIndexing.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringMethods/at.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringMethods/codePointAt.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringMethods/endsWith.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringMethods/fromCodePoint.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringMethods/includes.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringMethods/indexOf.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringMethods/lastIndexOf.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringMethods/padEnd.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringMethods/padStart.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringMethods/repeat.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringMethods/slice.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringMethods/split.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringMethods/startsWith.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringMethods/substring.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringMethods/toLowerCase.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringMethods/toUpperCase.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringMethods/trim.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringMethods/trimEnd.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/stringMethods/trimStart.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/swap.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/testBinaryTree.ts ...

Testing /Users/andrew/workspaces/ValueScript/inputs/passing/transitiveCapture.ts ...
test test_inputs::tests::test_inputs ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.94s
```
</details>

By comparison, this is the output on your branch:

<details>
<summary>Click to expand</summary>

```
➜  ValueScript git:(tarun_samanta/#7) cargo test -p vstc -- --nocapture
   Compiling valuescript_common v0.1.0 (/Users/andrew/workspaces/ValueScript/valuescript_common)
   Compiling valuescript_compiler v0.1.0 (/Users/andrew/workspaces/ValueScript/valuescript_compiler)
   Compiling valuescript_vm v0.1.0 (/Users/andrew/workspaces/ValueScript/valuescript_vm)
   Compiling vstc v0.1.0 (/Users/andrew/workspaces/ValueScript/vstc)
    Finished test [unoptimized + debuginfo] target(s) in 5.44s
     Running unittests src/main.rs (target/debug/deps/vstc-59fe0ddeb910d8de)

running 1 test
test test_inputs::tests::test_inputs ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s
```
</details>

**Comment by voltrevo** [comment]
Hi @tarunsamanta2k20, it looks like your last commit might be a mistake. Maybe you meant to commit something else? It's just added the compiled javascript files. My previous comment reviewing your changes still stands.

The compiled javascript files are also an additional problem. I'll need you to undo this commit before proceeding:

```sh
# Create a backup of your current branch
git checkout -b backup

# Switch back to your branch
git checkout -

# Undo the last commit, this is the same as "Git: Undo Last Commit" in vs code
git reset --soft HEAD^1

# You should now see the js files as staged uncommitted changes

# Unstage those js files
git reset # unstage

# Now remove the js files, you probably have other ways of doing this, but here's how you can do it with git
git clean -dfn . # Just prints what git would remove, check this first
git clean -df . # Actually remove those files

# Overwrite the remote, using -f to allow removing history
git push -f
```

If you want to run `tsc` to check the types, use `tsc --noEmit`. This will give you all the type checking without generating the js files.

In future, please check the diff of your PR to make sure the changes there are intended before asking for a review.

**Comment by voltrevo** [comment]
@tarunsamanta2k20 Hmm I'm going to want this change sooner or later and I'm afraid I'm going to end up dropping this and doing it myself.

How about a different approach - add me as a collaborator on your fork, that way I can push to your branch and we can work on this together.

Rather than fix up this branch, I'd have us start over. There's new test files that need to be included anyway.

I suggest you just make a new branch from latest main do the literal `// test_output! output` to `//! test_output(output)` in all the test files. Then I'll take care of updating `test_inputs.rs`, and we can get it merged in.

Wdyt?

**Comment by SamantaTarun** [comment]
@voltrevo I have added you as a collaborator to my  fork. Let me know if anything required from my side.

**Comment by voltrevo** [comment]
@tarunsamanta2k20 Great. Here's what I want you to do now:

1. Create a new branch that is up to date with voltrevo/main
2. Update the files in `inputs` to use the new `//! test_ouput(output)` pattern (110 files)
3. Commit and push
4. Open a new PR
5. Double check github shows the correct changes on the PR
6. Let me know and I'll take care of the rest :)

You can also reuse this PR by force-pushing the branch if you know how that works. Otherwise a new branch and a new PR is fine.

**Comment by SamantaTarun** [comment]
@voltrevo inside (output) should I put the output which is just next to it?

**Comment by SamantaTarun** [comment]
@tarunsamanta2k20 Like // ! test_ouput(1) or it should be like that?

//! test_ouput(output) 1
// (Should be 0.)

**Comment by voltrevo** [comment]
> @tarunsamanta2k20 Like // ! test_ouput(1) or it should be like that?
> 
> //! test_ouput(output) 1 // (Should be 0.)

First one 👍

**Comment by tarunsamanta2k20** [comment]
@voltrevo i am getting 75 results in main branch. have you modified anything?

image

**Comment by SamantaTarun** [comment]
@voltrevo give me some time, will make all those changes.

**Comment by SamantaTarun** [comment]
@voltrevo these three files are not present in your main branch. that's why i couldn't resolve merge conflicts.

 inputs/failing/assignmentEvalOrder.ts
inputs/failing/doublePreInc.ts
inputs/failing/stringIteration.ts

**Comment by voltrevo** [comment]
@tarunsamanta2k20 Awesome.

> that's why i couldn't resolve merge conflicts

All good. I'll fix it up and take it from here :).

**Review by voltrevo — APPROVED** [review]

**Comment by voltrevo** [comment]
Looks like there's an issue with the gh action when running against PRs. I'll look into fixing that up next time. It's not a true failure of this change.

**Comment by voltrevo** [comment]
@tarunsamanta2k20 Congrats on being the first outside contributor to ValueScript 🥳.

**Pull request merged**